code/part-two/client/source/Moji.jsx

```diff
@@ -2,73 +2,131 @@
 import React from 'react';
 import { Link } from 'react-router-dom';
 
-import { getMoji } from './services/requests';
+import { getMoji, getSires, submitPayloads } from './services/requests';
 import { parseDna } from './services/parse_dna';
 
 export class Moji extends React.Component {
   constructor(props) {
     super(props);
     this.state = {
       address: null,
+      isLoaded: false,
+      isOwner: false,
+      isSire: false,
       moji: null,
       mojiView: null
     };
+    this.selectSire = this.selectSire.bind(this);
   }
 
   static getDerivedStateFromProps(nextProps, prevState) {
-    console.log('GETDERIVEDSTATEFROMPROPS: <Moji />');
     const address = nextProps.match.params.address;
     return {
-      address
+      address,
+      isLoaded: false,
+      isOwner: false,
+      isSire: false,
+      moji: null,
+      mojiView: null
     };
   }
 
   componentDidMount() {
-    console.log('COMPONENTDIDMOUNT: <Moji />');
     this.fetchMoji(this.state.address);
   }
 
   componentDidUpdate(prevProps, prevState) {
-    console.log('COMPONENTDIDUPDATE: <Moji />');
     if (this.state.address !== prevState.address) {
       this.fetchMoji(this.state.address);
     }
   }
 
   fetchMoji(address) {
-    console.log('fetchMoji');
     return getMoji(address)
-      .then(moji => this.setState({
-        moji,
-        mojiView: parseDna(moji.dna).view
-      }))
       .catch(err => {
         console.error(`Fetch moji failed for ${address}`, err);
-        this.setState({ moji: null });
+      })
+      .then(moji => {
+        this.setState({
+          isOwner: moji.owner === this.props.publicKey,
+          moji,
+          mojiView: parseDna(moji.dna).view
+        });
+        return getSires(moji.owner);
+      })
+      .then(sire => {
+        const isSire = sire.address === this.state.address;
+        if (isSire) {
+          this.setState(prevState => ({
+            mojiView: prevState.mojiView + ' 🎩'
+          }));
+        }
+        this.setState({ isSire });
```

Comment:
Would it better to collect all necessary props in an arbitrary object first off and then set the state using that object. Just thinking if `isSire` is ever `true` then it will have to rerender the DOM twice. I don't know if it's really necessary or not but just a thought

Reply:
(Side note: I actually removed lines 59-63 as part of a future PR)

Agreed that it will re-render twice, but afaik (idk how to explicitly test this), it will only be a minor re-render (eg, only 1 new DOM element appearing), and if there were any performance issues, React is supposed to batch up multiple `this.setState`s and render just once anyway.

https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous
> React may batch multiple setState() calls into a single update for performance.